File: compiler/bench/tak.c
// expected: 7
#include <stdio.h>
long tak(long x, long y, long z) {
    if (y >= x) return z;
    return tak(tak(x-1, y, z), tak(y-1, z, x), tak(z-1, x, y));
}
int main() {
    printf("%ld\n", tak(18, 12, 6));
    return 0;
}

File: compiler/bench/tak.fs
\ expected: 7

: tak ( x y z -- result )
  over 2 pick > if
    2 pick 1- 2 pick 2 pick recurse
    3 pick 1- 3 pick 5 pick recurse
    3 pick 1- 5 pick 5 pick recurse
    recurse nip nip nip
  else nip nip then ;

: main 18 12 6 tak . cr ;
